Implement vectorized comparisons for all vector types on PPC

midronij opened the pull request:
This contribution includes the following changes: 
- Masked vector addition has been enabled for all vector types on PPC (this was necessary to allow for testing of the comparison operators, since there are no other vectorized masked operations on PPC at this time)
- All vector comparison IL opcodes (`vcmpeq`, `vcmpne`, `vcmpgt`, `vcmplt`, `vcmpge`, `vcmple`) have been enabled for all types on PPC
- Codegen sequences for all comparison operators have been implemented on PPC
- Binary encoder unit tests have been added for newly enabled PPC instructions

IBMJimmyk reviewed (changes requested):
This is everything I could find so far. Maybe some addition comments could be added to explicitly highlight how `lt`, `le` and `ge` versions of most vector comparison instructions don't exist. I didn't quite realize why some of these implementations were so roundabout until I saw it was because the intuitive single instruction solution didn't exist most of the time.

gita-omr commented on compiler/p/codegen/OMRCodeGenerator.cpp:
```diff
@@ -1785,7 +1785,7 @@ bool OMR::Power::CodeGenerator::getSupportsOpCodeForAutoSIMD(TR::CPU *cpu, TR::I
 
    if (cpu->isAtLeast(OMR_PROCESSOR_PPC_P8) &&
-       (opcode.getVectorOperation() == TR::vadd || opcode.getVectorOperation() == TR::vsub || opcode.getVectorOperation() == TR::vmul || opcode.getVectorOperation() == TR::vabs || opcode.getVectorOperation() == TR::vmin || opcode.getVectorOperation() == TR::vmax) &&
+       (opcode.getVectorOperation() == TR::vadd || opcode.getVectorOperation() == TR::vsub || opcode.getVectorOperation() == TR::vmul || opcode.getVectorOperation() == TR::vabs || opcode.getVectorOperation() == TR::vmin || opcode.getVectorOperation() == TR::vmax || opcode.getVectorOperation() == TR::vmadd) &&
```
I agree. Maybe we can use this PR to clean it up? Especially since all those opcodes already have labels in the switch below?

gita-omr commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -3093,7 +3186,7 @@ TR::Register *OMR::Power::TreeEvaluator::inlineVectorBinaryOp(TR::Node *node, TR
    if (masked)
       {
-      TR_ASSERT_FATAL(resReg->getKind() == TR_VSX_VECTOR, "Only masked VSX opcodes are currenlty supported\n");
+      TR_ASSERT_FATAL(resReg->getKind() == TR_VSX_VECTOR || resReg->getKind() == TR_VRF, "Only masked VSX opcodes are currenlty supported\n");
```
If we add TR_VRF, the text of the assertion does not make sense. I will try to recall why it was added in the first place.

gita-omr replied:
I think it was my mistake to add this ASSERT in some earlier PR and it can be removed. But a comment "Note: both VRF and VSX registers are supported by xxsel" would be helpful I think.

midronij commented:
I've addressed all the comments. In particular, regarding the helpers, I've collapsed them into a single function which hopefully makes things a little more concise and easy to read. If everything looks good I think we can start getting ready to merge these changes

IBMJimmyk commented on compiler/p/codegen/OMRTreeEvaluator.cpp:
```diff
@@ -1016,108 +1016,200 @@ OMR::Power::TreeEvaluator::m2vEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+                   "Only 128-bit vectors are supported %s", node->getDataType().toString());
+
+      TR::DataType elementType = node->getDataType().getVectorElementType();
+      bool instructionAvailable = cg->comp()->target().cpu.isAtLeast(OMR_PROCESSOR_PPC_P9); // VMX vector NE instructions are only available on P9 and up
```
I would rename this to something that references being a check for Power9 or later to be more descriptive. Also this whitespace in this method looks off to me.

IBMJimmyk reviewed (approved):
The code itself looks good to me. I only have minor suggestions on a variable name change and some whitespace fixes.

midronij commented:
@knn-k could you please review and, if everything looks alright, merge?

knn-k commented:
Jenkins build aix,plinux

knn-k commented:
@midronij There are some unit test failures.

```
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [m10 tests, listed below:
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCRecordFormSanityTest.checkFlippedBit/14, where GetParam() = (vcmpneh, vcmpneh., [ 00 00 00 01 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCRecordFormSanityTest.checkFlippedBit/15, where GetParam() = (vcmpnew, vcmpnew., [ 00 00 00 01 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCRecordFormSanityTest.checkFlippedBit/18, where GetParam() = (vcmpequd, vcmpequd., [ 00 00 00 01 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCRecordFormSanityTest.checkFlippedBit/22, where GetParam() = (vcmpgtsd, vcmpgtsd., [ 00 00 00 01 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCTrg1Src2EncodingTest.encode/66, where GetParam() = (vcmpgtsw, vr31, vr0, vr0, [ 13 e0 03 c7 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCTrg1Src2EncodingTest.encode/67, where GetParam() = (vcmpgtsw, vr0, vr31, vr0, [ 10 1f 03 c7 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCTrg1Src2EncodingTest.encode/68, where GetParam() = (vcmpgtsw, vr0, vr0, vr31, [ 10 00 fb c7 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCTrg1Src2EncodingTest.encodeRecordForm/66, where GetParam() = (vcmpgtsw, vr31, vr0, vr0, [ 13 e0 03 c7 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCTrg1Src2EncodingTest.encodeRecordForm/67, where GetParam() = (vcmpgtsw, vr0, vr31, vr0, [ 10 1f 03 c7 ])
[2022-11-30T00:57:49.527Z] 28: [0;31m[  FAILED  ] [mVMX/PPCTrg1Src2EncodingTest.encodeRecordForm/68, where GetParam() = (vcmpgtsw, vr0, vr0, vr31, [ 10 00 fb c7 ])
```

gita-omr commented:
Did not check binary encoding tests but otherwise LGTM.

midronij commented:
Binary encoder tests should be fixed now. @knn-k could you please re-run the CI builds and (if everything passes) merge?

knn-k commented:
Jenkins build aix,plinux

knn-k reviewed (approved):
LGTM.

knn-k commented:
The failure on x86-64 macOS is not related to this PR.
Merging.
